### PR TITLE
redownload pkgs when source hash verification fails

### PR DESCRIPTION
**Pull request by gbaz**
Resolves https://github.com/haskell/cabal/issues/7451

(possibly)

Adds a call, when checking hashes in project planning, to also verify the hashes of existing tarballs downloaded from secure repos to make sure things haven't "changed out" as the might in the case of a mutable head.hackage or the like.

* [x] Any changes that could be relevant to users [have been recorded in the changelog](https://github.com/haskell/cabal/blob/master/CONTRIBUTING.md#changelog).

### CODE REVIEW

File: cabal-install/src/Distribution/Client/FetchUtils.hs [hunk 1]
```diff
@@ -118,6 +122,30 @@ checkRepoTarballFetched repo pkgid = do
...
+           RepoSecure{} ->
+             repoContextWithSecureRepo repoCtxt repo $ \repoSecure ->
+                  Sec.withIndex repoSecure $ \callbacks -> (do
+                    fileInfo <- Sec.indexLookupFileInfo callbacks pkgid
+                    infoVerified <-  do
+                      sz <- Sec.FileLength . fromInteger <$> getFileSize file
+                      if sz /= Sec.fileInfoLength (Sec.trusted fileInfo)
+                      then return False
+                      else Sec.compareTrustedFileInfo (Sec.trusted fileInfo) <$> Sec.computeFileInfo (Sec.Path file :: Sec.Path Sec.Absolute)
+                    if infoVerified then return True else warn verbosity "whoops" >>  return False) -- TODO
```
Comment (Mikolaj) [inline cabal-install/src/Distribution/Client/FetchUtils.hs:145]: When you get around to this TODO, could you also make this more readable? The `do` in parens is hard to parse. E.g., 

```hs
Sec.withIndex repoSecure $ \callbacks -> do
  let verification = do
        fileInfo <- Sec.indexLookupFileInfo callbacks pkgid
        sz <- Sec.FileLength . fromInteger <$> getFileSize file
        if sz /= Sec.fileInfoLength (Sec.trusted fileInfo)
        then return False
        else Sec.compareTrustedFileInfo (Sec.trusted fileInfo) <$> Sec.computeFileInfo (Sec.Path file :: Sec.Path Sec.Absolute)
  infoVerified <- 
    verification
    `Sec.catchChecked` (\(e :: Sec.InvalidPackageException) -> warn verbosity (show e) >> return False)
    `Sec.catchChecked` (\(e :: Sec.VerificationError) -> warn verbosity (show e) >> return False)
  if infoVerified then return True else warn verbosity "whoops" >>  return False) -- TODO
```

which also changes the semantics slightly, probably resulting in a less puzzling warning if an exception occurs.

**Review by Mikolaj — COMMENTED** [review]
That looks like it may improve things. How to test it in CI?

**Comment by gbaz** [comment]
I don't think this is very CI testable. The best way I can think to test it is to fetch a package, then go into the cached downloads, mess with the downloaded package, and then attempt another action on it. I don't necessarily think adding a new test here makes much sense.

File: cabal-install/src/Distribution/Client/FetchUtils.hs (hunk 1 above)
Comment (gbaz) [reply cabal-install/src/Distribution/Client/FetchUtils.hs]: Thanks for noticing the TODO. I refactored and cleaned up, but could not factor out into a let block as suggested -- the checked exceptions mechanism in hackage-security is fiddly, and trying some obvious things still resulted in a situation where the typechecker yelled at me. I'm sure it could be made to work, but I think the code after the refactor is sufficiently more clear.

**Review by Mikolaj — APPROVED** [review]
Indeed, just realigning the `do` made it much easier to parse. Too bad about no smart enough tests to merit including in CI. Thank you.

File: changelog.d/pr-8500 [hunk 2]
```diff
@@ -0,0 +1,10 @@
+synopsis: Print a warning when assertions are enabled
```
Comment (Kleidukos) [inline changelog.d/pr-8500:1]: Shoudn't it be:

```suggestion
synopsis: Redownload pkgs when source hash verification fails
```
?

**Review by Kleidukos — CHANGES_REQUESTED** [review]

**Review by Kleidukos — APPROVED** [review]

File: cabal-install/src/Distribution/Client/FetchUtils.hs [hunk 3]
```diff
@@ -118,6 +122,34 @@ checkRepoTarballFetched repo pkgid = do
...
       else return Nothing
 
+verifyFetchedTarball :: Verbosity -> RepoContext -> Repo -> PackageId -> IO Bool
+verifyFetchedTarball verbosity repoCtxt repo pkgid =
+   let file = packageFile repo pkgid
+       handleError :: IO Bool -> IO Bool
+       handleError act = do
+         res <- Safe.try act
+         case res of
+           Left e -> warn verbosity ("Error verifying fetched tarball " ++ file ++ ", will redownload: " ++ show (e :: SomeException)) >> pure False
```
Comment (Bodigrim) [inline cabal-install/src/Distribution/Client/FetchUtils.hs:132]: Is this error message triggered if `file` simply does not exist?

Comment (Mikolaj) [reply cabal-install/src/Distribution/Client/FetchUtils.hs:132]: Yes, this function `verifyFetchedTarball` is called only in one place in the code, for all elements of `repoTarballPkgsWithMetadataUnvalidated`, which seems to contain ids of any packages that come from a secure repo, regardless of whether they were ever downloaded:

https://github.com/haskell/cabal/pull/8500/files#diff-2adae381a2eb2d8c804d5396694a431e1e3babd816a5d2408a62213cf774a893R929-R935

Comment (Mikolaj) [reply cabal-install/src/Distribution/Client/FetchUtils.hs:132]: A solution would be to move the partitioning wrt if a package is already downloaded (the block containing `checkRepoTarballFetched`) earlier and apply it to `allPkgLocations` instead of `repoTarballPkgsWithoutMetadata`. Only packages that are already downloaded would then be checked for `RepoSecure` and then verified. Does that make sense?

Comment (Mikolaj) [reply cabal-install/src/Distribution/Client/FetchUtils.hs:132]: Or, better, avoiding some filesystem IO, partition `repoTarballPkgsWithMetadataUnvalidated` wrt if a package is already downloaded.
